seL4 and linux kernel tracing supporting utils (#1)
tiiuae/sel4-trace-support: README.md

> # sel4-trace-support - a collection of programs to convert and merge seL4 tracebuffer to/with ftrace logs (WIP)

Reviewer: Put things after `-` to normal text.
```suggestion
# sel4-trace-support

A collection of programs to convert and merge seL4 tracebuffer to/with ftrace logs (WIP).
```

> ### Build

Reviewer: It is generally recommended to have blank lines before and after a title.

```suggestion
### Build

```

```
git clone git@github.com:tiiuae/sel4-trace-support.git
cd ./sel4-trace-support
mkdir build && cd ./build
cmake ..
make
```

### Usage

> #### sel4-extract - a tool to convert seL4 tracebuffer logs to ftrace campatible binary format

Reviewer: A typo fix, blank line etc.
```suggestion
#### sel4-extract

A tool for converting seL4 tracebuffer logs to ftrace compatible binary format.

```

   ```
   ./sel4-extract/sel4-extract [fname in] [fname out]
   ```

> #### sel4-ftrace-merge - a tool to merge converted seL4 tracebuffer logs with trace-cmd\kernelshart logs

Reviewer: Fix typos and new line etc.

```suggestion
#### sel4-ftrace-merge

A tool for merging converted seL4 tracebuffer logs with trace-cmd\kernelshark logs.

```

   ```
   ./sel4_trace_merge/sel4_trace_merge [ftrace_fname in] [sel4_fname in]
   ```

> ### Format description

Reviewer: Newline
```suggestion
### Format description

```

> trace-cmd binary format described at `man trace-cmd.dat` (also you can find this description in useful links)  

Reviewer: You could use [reference style links](https://www.markdownguide.org/basic-syntax/#reference-style-links)

```suggestion
trace-cmd binary format is described at `trace-cmd.dat` [man page][1] .

... text ...
... plaa ...

[1]: https://man7.org/linux/man-pages/man5/trace-cmd.dat.5.html
```

> ftrace binary described in following parts of trace-cmd.dat (**_Please be noticed: now this formats are hardcoded, but them should be actually parsed from following places_**):

Reviewer: ```suggestion
ftrace binary described in following parts of trace-cmd.dat (**_Please notice: currently formats are hardcoded, but they should be parsed from following places_**):
```

> + HEADER INFO FORMAT - contains a page (4096b) header format and event header formats, also event header format could be found here: `debugfs/tracing/events/header_event`

Reviewer: Perhaps

```suggestion
+ HEADER INFO FORMAT - contains a page (4096b) header format and event header formats. Event header format can also be found `debugfs/tracing/events/header_event`-file
```

   PAGE HEADER FORMAT example (debug output of `sel4-ftrace-merge` tool):
   ```
   [DBG]: ------------- Header Page Format -------------
   [DBG]: Header page format magic: header_page
   [DBG]: Header page format size: 205
   [DBG]: Page Header Format:
           field: u64 timestamp;   offset:0;       size:8; signed:0;
           field: local_t commit;  offset:8;       size:8; signed:1;
           field: int overwrite;   offset:8;       size:1; signed:1;
           field: char data;       offset:16;      size:4080;      signed:0;
   ```
   

>    Some format description:

Reviewer: ```suggestion
   Field descriptions:
```

   ```
      timestamp -- timestamp of a 1st entry in this page

>       commit    -- numer of bytes used in this page (up to 4080)

Reviewer: ```suggestion
      commit    -- number of bytes used in the page (up to 4080)
```

      overwrite -- don't know what is it for =(

>       data      -- data content of a page

Reviewer: ```suggestion
      data      -- data content of the page
```

   ```

   EVENT HEADER FORMAT example (debug output of `sel4-ftrace-merge` tool):
   ```
   [DBG]: ------------- Header Event Format -------------
   [DBG]: Header event magic: header_event
   [DBG]: Header event size: 205
   [DBG]: Header Event Format:
   # compressed entry header
           type_len    :    5 bits
           time_delta  :   27 bits
           array       :   32 bits

           padding     : type == 29
           time_extend : type == 30
           time_stamp : type == 31

>            data max type_len  == 28

Reviewer: differs from above... intentional?

Author: Sorry, I probably did not get you, but:
```
time_delta -- time difference with timestamp in PAGE HEADER (for 1st entry is zero)
```
time_delta is difference btwn timestamp in PAGE HEADER and time in current event. Is this sounds like an answer for your question?

Reviewer: Nevermind :)

   ```

>    Some format description:

Reviewer: ```suggestion
   Field description:
```

         this one is only one actually useful and well described case
         (even in `trace-cmd` code, so other cases probably **dead**)
   ```
   
+ FTRACE EVENT FORMATS - contains an events format descriptions, also could be found here: `debugfs/tracing/events/ftrace/<event>/format`

   We need `FUNCTION` event, here is an example (debug output of `sel4-ftrace-merge` tool):
   ```[DBG]:
   [DBG]: ------------- Ftrace Event Formats -------------
   [DBG]: Ftrace event formats count: 15
   .....
   [DBG]: Event Format #8 size: 0x1cc
   [DBG]: Event Format:
   name: function
   ID: 1
   format:
           field:unsigned short common_type;       offset:0;       size:2; signed:0;
           field:unsigned char common_flags;       offset:2;       size:1; signed:0;
           field:unsigned char common_preempt_count;       offset:3;       size:1; signed:0;
           field:int common_pid;   offset:4;       size:4; signed:1;

           field:unsigned long ip; offset:8;       size:8; signed:0;
           field:unsigned long parent_ip;  offset:16;      size:8; signed:0;

   print fmt: " %ps <-- %ps", (void *)REC->ip, (void *)REC->parent_ip
   ```

>    Some format description:

Reviewer: Leading and trailing newline for clarity.
```suggestion

   Field descriptions:
   
```

      common_pid           -- linux process ID, for seL4 we use here some dedicated PID (like: 999, also hardcoded =)
      ip                   -- function IP, for for seL4 we use here some dedicated IP ( also hardcoded =)
      parent_ip            -- parent function IP (caller), for for seL4 we use here some dedicated IP ( also hardcoded =)
   ```
   
+ PAGE FORMAT example:
   ```
   PAGE_HEADER {
      ...
      data {
        EVENT_HEADER {
           ...
           array {
              EVENT_FORMAT {}
           }
        },
        EVENT_HEADER {
           ...
           array {
              EVENT_FORMAT {}
           }
        },
        EVENT_HEADER {
           ...
           array {
              EVENT_FORMAT {}
           }
        },
        ...
      }
   }
   ```

+ PROCESS INFORMATION   -- section contains current processes information from: `debugfs/tracing/saved_cmdlines`

   Here we need to add our seL4 `common_pid` and it's name

+ KALLSYMS INFORMATION  -- section contains current kernel address to function maps from: `/proc/kallsyms`

   Here we need to add our seL4 `ip`, `parent_ip` and their names

### Useful links

> [man trace-cmd](https://man7.org/linux/man-pages/man5/trace-cmd.dat.5.html)  

Reviewer: As mentioned above, some of these could be reference style links, and some could be like they currently are. Your call :)

Author: Sure, thank you for all your suggestions. But here I think that there is not much difference to use this approach or any other one. So keep them as is.

Reviewer: Alright, np
